bindings/matrix-sdk-ffi/src/widget.rs

```diff
@@ -26,23 +26,175 @@ impl From<Widget> for matrix_sdk::widget::Widget {
+    fn from(value: matrix_sdk::widget::WidgetSettings) -> Self {
+        let matrix_sdk::widget::WidgetSettings { id, init_after_content_load, raw_url } = value;
+        WidgetSettings { id, init_after_content_load: init_after_content_load, raw_url }
+    }
+}
+
+#[derive(Debug, thiserror::Error, uniffi::Error)]
+#[uniffi(flat_error)]
+pub enum ParseError {
+    #[error("empty host")]
+    EmptyHost,
+    #[error("invalid international domain name")]
+    IdnaError,
+    #[error("invalid port number")]
+    InvalidPort,
+    #[error("invalid IPv4 address")]
+    InvalidIpv4Address,
+    #[error("invalid IPv6 address")]
+    InvalidIpv6Address,
+    #[error("invalid domain character")]
+    InvalidDomainCharacter,
+    #[error("relative URL without a base")]
+    RelativeUrlWithoutBase,
+    #[error("relative URL with a cannot-be-a-base base")]
+    RelativeUrlWithCannotBeABaseBase,
+    #[error("a cannot-be-a-base URL doesn’t have a host to set")]
+    SetHostOnCannotBeABaseUrl,
+    #[error("URLs more than 4 GB are not supported")]
+    Overflow,
+    #[error("unknown parse error")]
+    Other,
+}
+
+impl From<url::ParseError> for ParseError {
+    fn from(value: url::ParseError) -> Self {
+        match value {
+            url::ParseError::EmptyHost => Self::EmptyHost,
+            url::ParseError::IdnaError => Self::IdnaError,
+            url::ParseError::InvalidPort => Self::InvalidPort,
+            url::ParseError::InvalidIpv4Address => Self::InvalidIpv4Address,
+            url::ParseError::InvalidIpv6Address => Self::InvalidIpv6Address,
+            url::ParseError::InvalidDomainCharacter => Self::InvalidDomainCharacter,
+            url::ParseError::RelativeUrlWithoutBase => Self::RelativeUrlWithoutBase,
+            url::ParseError::RelativeUrlWithCannotBeABaseBase => {
+                Self::RelativeUrlWithCannotBeABaseBase
+            }
+            url::ParseError::SetHostOnCannotBeABaseUrl => Self::SetHostOnCannotBeABaseUrl,
+            url::ParseError::Overflow => Self::Overflow,
+            _ => Self::Other,
+        }
     }
 }
 
+/// Create the actual url that can be used to setup the WebView or IFrame
+/// that contains the widget.
+///
+/// # Arguments
+/// * `widget_settings` - The widget settings to generate the url for.
+/// * `room` - A matrix room which is used to query the logged in username
+/// * `props` - Properties from the client that can be used by a widget to adapt
+///   to the client. e.g. language, font-scale...
+#[uniffi::export]
+pub async fn generate_url(
+    widget_settings: WidgetSettings,
+    room: Arc<Room>,
+    props: ClientProperties,
+) -> Result<String, ParseError> {
+    Ok(matrix_sdk::widget::WidgetSettings::generate_url(
+        &widget_settings.clone().into(),
+        &room.inner,
+        props.into(),
+    )
+    .await
+    .map(|url| url.to_string())?)
+}
+/// `WidgetSettings` are usually created from a state event.
+/// (currently unimplemented)
```

Comment:
It seems implemented, no?

Reply:
This implementation generates `WidgetSettings` from fn params. Most widgets will be part of the room state.
What is not yet implemented is will be `room.get_widgets() -> Vec<WidgetSettings>` (where `room: Room`.)